tests/beta/interfaces/test_tape_autograd.py

```diff
@@ -0,0 +1,668 @@
+
+        dev = qml.device("default.qubit.autograd", wires=1)
+
+        res = cost_fn(a, p, device=dev)
+        expected = np.cos(a) * np.cos(p[1]) * np.sin(p[0]) + np.sin(a) * (
+            np.cos(p[2]) * np.sin(p[1]) + np.cos(p[0]) * np.cos(p[1]) * np.sin(p[2])
+        )
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        jac_fn = qml.jacobian(cost_fn)
+
+        res = jac_fn(a, p, device=dev)
+        expected = np.array(
+            [
+                np.cos(p[1]) * (np.cos(a) * np.cos(p[0]) - np.sin(a) * np.sin(p[0]) * np.sin(p[2])),
+                np.cos(p[1]) * np.cos(p[2]) * np.sin(a)
+                - np.sin(p[1])
+                * (np.cos(a) * np.sin(p[0]) + np.cos(p[0]) * np.sin(a) * np.sin(p[2])),
+                np.sin(a)
+                * (np.cos(p[0]) * np.cos(p[1]) * np.cos(p[2]) - np.sin(p[1]) * np.sin(p[2])),
+            ]
+        )
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+        spy.assert_not_called()
+
+    def test_probability_differentiation(self, mocker, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        spy = mocker.spy(QuantumTape, "jacobian")
+
+        def cost(x, y, device):
+            with QuantumTape() as tape:
+                qml.RX(x, wires=[0])
+                qml.RY(y, wires=[1])
+                qml.CNOT(wires=[0, 1])
+                probs(wires=[0])
+                probs(wires=[1])
+
+            return tape.execute(device)
+
+        dev = qml.device("default.qubit.autograd", wires=2)
+        x = np.array(0.543, requires_grad=True)
+        y = np.array(-0.654, requires_grad=True)
+
+        res = cost(x, y, device=dev)
+        expected = np.array(
+            [
+                [np.cos(x / 2) ** 2, np.sin(x / 2) ** 2],
+                [(1 + np.cos(x) * np.cos(y)) / 2, (1 - np.cos(x) * np.cos(y)) / 2],
+            ]
+        )
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        jac_fn = qml.jacobian(cost)
+        res = jac_fn(x, y, device=dev)
+        assert res.shape == (2, 2, 2)
+
+        expected = np.array(
+            [
+                [[-np.sin(x) / 2, 0], [-np.sin(x) * np.cos(y) / 2, -np.cos(x) * np.sin(y) / 2]],
+                [
+                    [np.sin(x) / 2, 0],
+                    [np.cos(y) * np.sin(x) / 2, np.cos(x) * np.sin(y) / 2],
+                ],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+        spy.assert_not_called()
+
+    def test_ragged_differentiation(self, mocker, monkeypatch, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        spy = mocker.spy(QuantumTape, "jacobian")
+        dev = qml.device("default.qubit.autograd", wires=2)
+
+        def _asarray(args, dtype=np.float64):
+            return np.hstack(args).flatten()
+
+        # we need to patch the asarray method on the device

```

Comment:
how come?

Reply:
See https://github.com/PennyLaneAI/pennylane/pull/804/files#r489828719

Reply:
lol that was meant as a prompt to put the explanation in the comment :laughing:

Reply:
will add 😆